crates/bevy_input_focus/src/directional_navigation.rs

```diff
@@ -259,8 +469,317 @@ pub enum DirectionalNavigationError {
+/// containing its entity ID, center position, and size for spatial navigation calculations.
+///
+/// The term "focusable area" avoids confusion with UI [`Node`](bevy_ui::Node) components.
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(Reflect),
+    reflect(Debug, PartialEq, Clone)
+)]
+pub struct FocusableArea {
+    /// The entity identifier for this focusable area.
+    pub entity: Entity,
+    /// The center position in global coordinates.
+    pub position: Vec2,
+    /// The size (width, height) of the area.
+    pub size: Vec2,
+}
+
+/// Trait for extracting position and size from navigable UI components.
+///
+/// This allows the auto-navigation system to work with different UI implementations
+/// as long as they can provide position and size information.
+pub trait Navigable {
+    /// Returns the center position and size in global coordinates.
+    fn get_bounds(&self) -> (Vec2, Vec2);
+}
+
+// We can't directly implement this for `bevy_ui` types here without circular dependencies,
+// so we'll use a more generic approach with separate functions for different component sets.
+
+/// Calculate the overlap factor between two nodes in the perpendicular axis.
+///
+/// Returns a value between 0.0 (no overlap) and 1.0 (perfect overlap).
+/// For diagonal directions, always returns 1.0.
+fn calculate_overlap(
+    origin_pos: Vec2,
+    origin_size: Vec2,
+    candidate_pos: Vec2,
+    candidate_size: Vec2,
+    octant: CompassOctant,
+) -> f32 {
+    match octant {
+        CompassOctant::North | CompassOctant::South => {
+            // Check horizontal overlap
+            let origin_left = origin_pos.x - origin_size.x / 2.0;
+            let origin_right = origin_pos.x + origin_size.x / 2.0;
+            let cand_left = candidate_pos.x - candidate_size.x / 2.0;
+            let cand_right = candidate_pos.x + candidate_size.x / 2.0;
+
+            let overlap = (origin_right.min(cand_right) - origin_left.max(cand_left)).max(0.0);
+            let max_overlap = origin_size.x.min(candidate_size.x);
+            if max_overlap > 0.0 {
+                overlap / max_overlap
+            } else {
+                0.0
+            }
+        }
+        CompassOctant::East | CompassOctant::West => {
+            // Check vertical overlap
+            let origin_bottom = origin_pos.y - origin_size.y / 2.0;
+            let origin_top = origin_pos.y + origin_size.y / 2.0;
+            let cand_bottom = candidate_pos.y - candidate_size.y / 2.0;
+            let cand_top = candidate_pos.y + candidate_size.y / 2.0;
+
+            let overlap = (origin_top.min(cand_top) - origin_bottom.max(cand_bottom)).max(0.0);
+            let max_overlap = origin_size.y.min(candidate_size.y);
+            if max_overlap > 0.0 {
+                overlap / max_overlap
+            } else {
+                0.0
+            }
+        }
+        // Diagonal directions don't require strict overlap
+        _ => 1.0,
```

Comment:
For diagonals we could project onto the diagonal axis and the compare the overlap, or is that not necessary?